### PR TITLE
tempfile.SpooledTemporaryFile: inherit from IOBase on 3.11

### PR DESCRIPTION
**Pull request by JelleZijlstra**
python/cpython#29560

### CODE REVIEW

**Review by hauntsaninja — APPROVED** [review]

**Review by AlexWaygood — COMMENTED** [review]
I'd prefer us to use a wordier-but-more-readable alias name, like `_SpooledTemporaryFileBase`, rather than `_STFBase` which feels a bit opaque.

File: stdlib/tempfile.pyi [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import io
 import os
 import sys
 from _typeshed import Self
```
Comment (AlexWaygood) [inline stdlib/tempfile.pyi:4]:
```suggestion
from _typeshed import Self, WriteableBuffer
```

File: stdlib/tempfile.pyi [hunk 2]
```diff
@@ -318,6 +324,13 @@ class SpooledTemporaryFile(IO[AnyStr]):
     def fileno(self) -> int: ...
     def flush(self) -> None: ...
     def isatty(self) -> bool: ...
+    if sys.version_info >= (3, 11):
+        def readable(self) -> bool: ...
```
Comment (AlexWaygood) [inline stdlib/tempfile.pyi:328]:
```suggestion
```

File: stdlib/tempfile.pyi [hunk 3]
```diff
@@ -318,6 +324,13 @@ class SpooledTemporaryFile(IO[AnyStr]):
...
+        def read1(self, __size: int = ...) -> AnyStr: ...
+        def readinto(self, b: WriteableBuffer) -> int: ...
+        def readinto1(self, b: WriteableBuffer) -> int: ...
+        def seekable(self) -> bool: ...
```
Comment (AlexWaygood) [inline stdlib/tempfile.pyi:332]:
```suggestion
```

**Review by AlexWaygood — COMMENTED** [review]
In typeshed we've always pretended that `SpooledTemporaryFile` has `readable()` and `seekable()` methods on <3.11, even though it doesn't at runtime, because it inherits from `IO` in typeshed, and `IO` has these as abstract methods.

File: stdlib/tempfile.pyi (hunk 2 above)
Comment (JelleZijlstra) [reply stdlib/tempfile.pyi]: Thanks, applied this and updated the comments.

**Review by AlexWaygood — COMMENTED** [review]
What about the new `detach()` method? https://github.com/python/cpython/pull/29560/files#diff-6553a99f3ae04c9fc9c2349ac27037bfe3b274d5ca54d5800d9c5b5f11e29d21R813

File: stdlib/tempfile.pyi [hunk 4]
```diff
@@ -318,20 +324,26 @@ class SpooledTemporaryFile(IO[AnyStr]):
     def fileno(self) -> int: ...
     def flush(self) -> None: ...
     def isatty(self) -> bool: ...
+    if sys.version_info >= (3, 11):
+        def read1(self, __size: int = ...) -> AnyStr: ...
+        def readinto(self, b: WriteableBuffer) -> int: ...
+        def readinto1(self, b: WriteableBuffer) -> int: ...
```
Comment (AlexWaygood) [inline stdlib/tempfile.pyi:330]: I'm confused by these methods.

According to the runtime, these call `self._file.read1()`, `self._file.readinto()`, `self._file.readinto1()`: https://github.com/python/cpython/pull/29560/files#diff-6553a99f3ae04c9fc9c2349ac27037bfe3b274d5ca54d5800d9c5b5f11e29d21R766-R774

But `self._file` could be a `BytesIO` or a `TextIOWrapper`: https://github.com/python/cpython/pull/29560/files#diff-6553a99f3ae04c9fc9c2349ac27037bfe3b274d5ca54d5800d9c5b5f11e29d21R653-R656

And according to our current stubs, `TextIOWrapper` doesn't have a `read1`, `readinto` or `readinto1` method: https://mypy-play.net/?mypy=latest&python=3.10&gist=c5b47d63657de3a0a8c9a275baf16705

Comment (Akuli) [reply stdlib/tempfile.pyi:330]: Whether or not these exist seems to depend on the mode:

```
>>> tempfile.TemporaryFile("r").read1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: '_io.TextIOWrapper' object has no attribute 'read1'
>>> tempfile.TemporaryFile("rb").read1
<built-in method read1 of _io.BufferedReader object at 0x7fb6c9dd67d0>
```

I can't think of a practical use case for `read1()` or `readinto1()`. With subprocesses, `proc.stdout.read1(100)` reads data from the stdout of a subprocess without waiting for more data to arrive (#6153). But with tempfiles, a plain `.read()` works like that and just returns empty bytes once it's at the end of the file.

File: stdlib/tempfile.pyi [hunk 5]
```diff
@@ -318,20 +324,26 @@ class SpooledTemporaryFile(IO[AnyStr]):
...
+        def detach(self) -> io.RawIOBase: ...
+
     def read(self, n: int = ...) -> AnyStr: ...
-    def readline(self, limit: int = ...) -> AnyStr: ...
-    def readlines(self, hint: int = ...) -> list[AnyStr]: ...
+    def readline(self, limit: int | None = ...) -> AnyStr: ...  # type: ignore[override]
+    def readlines(self, hint: int = ...) -> list[AnyStr]: ...  # type: ignore[override]
```
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply stdlib/tempfile.pyi:335]: I'll mark them as positional-only to match the `*args` declaration.

File: stdlib/tempfile.pyi (hunk 4 above)
Comment (JelleZijlstra) [reply stdlib/tempfile.pyi:330]: Adding a comment.

File: stdlib/tempfile.pyi [hunk 6]
```diff
@@ -217,9 +218,14 @@ class _TemporaryFileWrapper(Generic[AnyStr], IO[AnyStr]):
     def write(self, s: AnyStr) -> int: ...
     def writelines(self, lines: Iterable[AnyStr]) -> None: ...
 
-# It does not actually derive from IO[AnyStr], but it does implement the
-# protocol.
-class SpooledTemporaryFile(IO[AnyStr]):
+# It does not actually derive from IO[AnyStr], but it does mostly behave
+# like one.
```
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply stdlib/tempfile.pyi:222]: Done

**Review by AlexWaygood — APPROVED** [review]

**Comment by JelleZijlstra** [comment]
hm, I expected unused stubtest allowlist entries. I'll look into that in a followup PR.

**Pull request merged**